File: Programas/NoInformados/Rompecabezas/2. Busqueda en Amplitud/PYTHON BFS/BFS.py
from collections import deque

def trayectoria( nodoFinal ):
    secuencia = deque()
    secuencia.append( nodoFinal )
    while nodoFinal.nodoPadre != None :
        nodoFinal = nodoFinal.nodoPadre
        secuencia.append( nodoFinal )
    secuencia.reverse()
    return secuencia

class BFS:
    
    @staticmethod
    def Buscar( nodoInicial , nodoFinal ):
        agenda = deque()
        configuracionesExploradas = set()

        if( nodoFinal.compararCeldas(nodoInicial) ):
            return trayectoria(nodoInicial)

        agenda.append( nodoInicial )

        while( agenda ):
            nodo = agenda.popleft()

            configuracionesExploradas.add( nodo.tablero )

            for hijo in nodo.expandirNodos():
                if( nodoFinal.compararCeldas(hijo) ):
                    return trayectoria(hijo)
                elif( not(hijo.tablero in configuracionesExploradas) ):
                    agenda.append( hijo )
        return None

File: Programas/NoInformados/Rompecabezas/2. Busqueda en Amplitud/JAVA BFS/BFS.java
import java.util.ArrayList;
import java.util.Collections;
import java.util.HashSet;

public class BFS {

    // Recupera los nodos visitados para llegar al nodo final
    private static ArrayList<Rompecabezas> Trayectoria( Rompecabezas nodoFinal ){
        ArrayList<Rompecabezas> secuencia = new ArrayList<>();
        secuencia.add( nodoFinal );
        while( nodoFinal.nodoPadre != null ){
            nodoFinal = nodoFinal.nodoPadre;
            secuencia.add( nodoFinal );
        }
        Collections.reverse(secuencia);
        return secuencia;
    }

    // Algoritmo Busqueda en Profundidad
    public static ArrayList<Rompecabezas> Buscar( Rompecabezas nodoInicial , Rompecabezas nodoFinal ){
        ArrayList<Rompecabezas> agenda = new ArrayList<>();
        HashSet<Long> configuracionesExploradas = new HashSet<>();

        if( nodoFinal.compararCeldas(nodoInicial) )
            return Trayectoria(nodoInicial);
        
        agenda.add( nodoInicial );

        while( agenda.size() > 0 ){
            Rompecabezas nodo = agenda.remove( 0 );

            configuracionesExploradas.add( nodo.tablero );
            for( Rompecabezas hijo : nodo.expandirNodos() ){
                if( nodoFinal.compararCeldas(hijo) )
                    return Trayectoria( hijo );
                else if( !configuracionesExploradas.contains( hijo.tablero ) )
                    agenda.add( hijo );
            }
        }

        return null;
    }
}
